src/PersistentOrderedMap.mo

```diff
@@ -0,0 +1,925 @@
+/// Stable key-value map implemented as a red-black tree with nodes storing key-value pairs.
+///
+/// A red-black tree is a balanced binary search tree ordered by the keys.
+///
+/// The tree data structure internally colors each of its nodes either red or black,
+/// and uses this information to balance the tree during the modifying operations.
+///
+/// Performance:
+/// * Runtime: `O(log(n))` worst case cost per insertion, removal, and retrieval operation.
+/// * Space: `O(n)` for storing the entire tree.
+/// `n` denotes the number of key-value entries (i.e. nodes) stored in the tree.
+///
+/// Note:
+/// * Map operations, such as retrieval, insertion, and removal create `O(log(n))` temporary objects that become garbage.
+///
+/// Credits:
+///
+/// The core of this implementation is derived from:
+///
+/// * Ken Friis Larsen's [RedBlackMap.sml](https://github.com/kfl/mosml/blob/master/src/mosmllib/Redblackmap.sml), which itself is based on:
+/// * Stefan Kahrs, "Red-black trees with types", Journal of Functional Programming, 11(4): 425-432 (2001), [version 1 in web appendix](http://www.cs.ukc.ac.uk/people/staff/smk/redblack/rb.html).
+
+
+import Debug "Debug";
+import I "Iter";
+import List "List";
+import Nat "Nat";
+import O "Order";
+
+// TODO: a faster, more compact and less indirect representation would be:
+// type Map<K, V> = {
+//  #red : (Map<K, V>, K, V, Map<K, V>);
+//  #black : (Map<K, V>, K, V, Map<K, V>);
+//  #leaf
+//};
+// (this inlines the colors into the variant, flattens a tuple, and removes a (now) redundant option, for considerable heap savings.)
```

Comment:
One more optimation to try is flattening the (K,V) tuple (as suggested in the comment). That would save space, allocation and remove an indirection. I think there might be a decent win there.

Reply:
Ah, I see you already tried that here: https://github.com/serokell/motoko-base/pull/15
Maybe not worth it then...

Reply:
I guess its not necessarily an improvement because the rebalancing operations can share the tuples but not the flattened fields?

Reply:
Ah, but looking at the max memory numbers in https://github.com/serokell/motoko-base/pull/15 (which in my reading go down 20%)  perhaps  it is worth it...?

Reply:
What do you think? Maybe @luc-blaeser has an opinion...

Reply:
I don't have a strong opinion on this. Maybe the saving could be worth doing. I guess balancing is not becoming more expensive as the key and values can directly shared in the new (flat) node object.